### PR TITLE
Node.js hometask

### PR DESCRIPTION
**Pull request by oksanaperun**

### CODE REVIEW

File: backend/models/group.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+const mongoose = require('mongoose');
+const Schema = mongoose.Schema;
+
+const groupSchema = new Schema({
+    name: String,
+    state: Boolean
```
Comment (nakhodkin) [inline backend/models/group.js:6]: _Trivial_
You could specify here an additional field for devices, attached to this group e.g.
`devices: [{ type: Schema.Types.ObjectId, ref: 'Device' }]`

However, it is not necessary in the case of this app

File: backend/models/group.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+});
+
+const groupModel = mongoose.model('group', groupSchema);
```
Comment (nakhodkin) [inline backend/models/group.js:9]: _trivial_
You can try to define the model with a capital letter, don't be afraid, that your collections might break because it becomes `Groups`. Actually, they don't 😊, mongoose will do lowercase + define collection name in a plural form of the model name by itself.

File: backend/routes/devices.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        await device.updateOne({
+            ...deviceData,
+            state: isStateChange ? !device.state : device.state
+        });
+
+        if (isStateChange) {
+            await utils.logAction(deviceId, deviceData.state, false);
+            await utils.toggleRealDevice(device);
+        }
+
```
Comment (nakhodkin) [inline backend/routes/devices.js:58]: `trivial`
Given that, you might want to define `devices` as an additional field for the `Group` you might need some additional logic of managing the data under this field e.g.

```javascript
        if (groupId) {
            const group = await Group.findById(groupId).exec();
            await group.updateOne({
                devices: [...group.devices, deviceId]
            });
        }

        if (!groupId) {
            const [ group ] = await Group.find({ devices: deviceId }).exec();
            await group.updateOne({
                devices: group.devices.filter((id) => id != deviceId)
            });
        }
```

PS this `devices: group.devices.filter((id) => id != deviceId)` is not as good practice as normally the data is coming from the frontend in plain strings, but here comparison between `String` and `ObjectId` happens. `==` will work, because it will coerce `ObjectId` to `String`, however, it might be better to convert it explicitly.

File: backend/routes/devices.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        res.sendStatus(200);
+    } catch (e) {
+        res.sendStatus(404);
```
Comment (nakhodkin) [inline backend/routes/devices.js:61]: `trivial`
This Handler comprises a lot of business logic. More likely that `500` has better chances to take place here😁. But you never know. Probably in the most simple case, it would be not that bad to send the stringified exception back to the client e.g.

`res.status(404).send(e);`

File: backend/routes/devices.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    }
+});
+
+function deviceAdapter(device) {
```
Comment (nakhodkin) [inline backend/routes/devices.js:65]: `trivial`
Wow. This is actually an awesome decision to adapt the data format before sending it to the client 👍🏻. Probably this stuff is known better as `DTO's` I'm not sure.

File: frontend/src/scenes/Groups.jsx [hunk 6]
```diff
@@ -1,42 +1,61 @@
...
+                    </div>
                 </div>
-            </div>
 
-            <div className="row mt-2">
-                <div className="col-3 offset-9">
-                    <a href="#/groups/add" className="btn btn-primary float-right">Add group</a>
+                <div className="row mt-2">
+                    <div className="col-3 offset-9">
+                        <a href="#/groups/add" className="btn btn-primary float-right">Add group</a>
```
Comment (nakhodkin) [inline frontend/src/scenes/Groups.jsx:36]: `trivial`
You might want to use `<Link />` component from `react-router-dom` since you've already been using this package for routing in your app.

File: frontend/src/constants.js [hunk 7]
```diff
@@ -5,5 +5,11 @@ export const devicePropType = PropTypes.shape({
     name: PropTypes.string.isRequired,
     address: PropTypes.string.isRequired,
     port: PropTypes.number.isRequired,
-    state: PropTypes.oneOf(['on', 'off'])
-});
\ No newline at end of file
+    state: PropTypes.oneOf(['On', 'Off'])
+});
+
+export const groupPropType = PropTypes.shape({
```
Comment (nakhodkin) [inline frontend/src/constants.js:11]: `trivial`
This is actually not bad, that you're avoiding code duplication by bringing this out to the constants.
However, here is a tricky part... Ok, you've defined the `GropType` which has the same meaning against all the code base of the frontend, but the gotcha is, that components in your application might have very different interfaces and it the thinner they are - the better.

Like here is an example of a simple component, which brings the fancy styling for a switch toggle which displays the status of the group:
```javascript
import { groupPropType } from './constants';

const SwitchToggle = ({ status }) => <div style={/* fancy styling */} />
SwitchToggle.propTypes = groupPropType;
```

Regardless, this is a sort of contrived example, but in a larger app you might have faced with an issues like - your `constants.js` might overgrow, or would be difficult to keep he propTypes definitions separately from the components up to date.

To summarize, this `propTypes` stuff is a sort of a meta-information, moreover, it is cut off from the prod bundle, so there is no propblem to not introduce `contanstats.js` for this purpose, but it's up to you. In your project it makes sense to me:)
